Taproot: BIP32 extended keys using Scep256k1 keys instead of bitcoin ECDSA

dr-orlovsky opened the pull request:
This is third step required to introduce Schnorr key support according to #588. This PR starts API-breaking changes and is follow-up to non-API breaking #589, which is already merged.

PR rationale: BIP32 does not support uncompressed keys and using type with compression flag was a mistake

thomaseizinger commented on src/util/bip32.rs:
```diff
@@ -502,13 +503,7 @@ impl ExtendedPrivKey {
-                    &hmac_result[..32]
-                ).map_err(Error::Ecdsa)?,
-            },
+            private_key: secp256k1::SecretKey::from_slice(&hmac_result[..32]).map_err(Error::Ecdsa)?,
```
We have a `From<secp256k1::Error> for Error` implementation so I think you should be able to get rid of the `map_err`:

```suggestion
            private_key: secp256k1::SecretKey::from_slice(&hmac_result[..32])?,
```

dr-orlovsky replied:
Fixed (in many other places as well)

thomaseizinger commented on src/util/bip32.rs:
```diff
@@ -619,6 +634,20 @@ impl ExtendedPubKey {
         }
     }
 
+    /// Constructs ECDSA compressed private key matching internal public key representation.
```
Small typo:

```suggestion
    /// Constructs ECDSA compressed public key matching internal public key representation.
```

thomaseizinger reviewed (commented):
ACK d155115a37618ac4e6e2924d445a453b40ee7c17

dr-orlovsky commented:
Rebased on master, addressed comments, now ready for review

sanket1729 commented on src/util/bip32.rs:
```diff
@@ -504,6 +504,21 @@ impl ExtendedPrivKey {
+
+    /// Constructs BIP340 keypair for Schnorr signatures and Taproot use matching the internal
+    /// secret key representation.
+    pub fn to_schnorr<C: secp256k1::Signing>(&self, secp: &Secp256k1<C>) -> schnorr::KeyPair {
```
I have a slight preference in favor of `to_schnorr_keypair`. 

I also like `to_ecdsa_privkey`, `to_ecdsa_pubkey`, and `to_schnorr_pubkey` over the current ones. But this one is more debatable since the user has the context of what return type to expect.

dr-orlovsky replied:
Changed to `to_keypair` since there is no other keypair (and @apoelstra does not like "Schnorr" surname :D because of him patenting signature algo)

sanket1729 replied:
I would prefer `to_schnorr_keypair`. It's sad that this has patent history, but IMO the convenience and usability triumphs this.

dr-orlovsky replied:
Do we have a keypair type for non-BIP-340 keys?

dr-orlovsky replied:
I really do not like inconsistency in function names for extended keys that we have today in the API: `derive_priv`, `derive_pub`, `new_master`, `from_private` - and now `to_ecdsa_privkey`.... :(

dr-orlovsky replied:
Did a temporal version with `priv/pub` instead of `privkey` for more consisted naming for now; will wait for others to comment

sanket1729 replied:
Reviewing the discussion, after some thinking I think my vote is for `to_ec_pubkey`, `to_xonly_keypair` for rationale mentioned here: 
https://github.com/rust-bitcoin/rust-bitcoin/issues/588#issuecomment-890586744

sanket1729 commented on src/util/bip32.rs:
```diff
@@ -615,6 +630,20 @@ impl ExtendedPubKey {
+        }
+    }
+
+    /// Constructs BIP340 xcoord-only public key for Schnorr signatures and Taproot use matching
```
Let's stick with the term `X-only` instead of `x-coord-only` to be more consistent with bip340

dr-orlovsky replied:
Last time I read it it was "X-coord only", but anyway, changed. Omnia mutantur...

sanket1729 reviewed (commented):
Thanks for this. The code looks great. I have a discussion point about API naming which might use more feedback from other contributors.

sanket1729 commented on src/util/bip32.rs:
```diff
@@ -638,9 +638,9 @@ impl ExtendedPubKey {
+    /// Constructs BIP340 x-only public key for BIP-340 signatures and Taproot use matching
     /// the internal public key representation.
-    pub fn to_schnorr(&self) -> schnorr::PublicKey {
+    pub fn to_keypair(&self) -> schnorr::PublicKey {
```
This returns a public key and not a kaypair

sgeisler commented on src/util/bip32.rs:
```diff
@@ -498,11 +499,26 @@ impl ExtendedPrivKey {
     }
 
+    /// Constructs ECDSA compressed private key matching internal secret key representation.
+    pub fn to_ecdsa_priv(&self) -> ecdsa::PrivateKey {
```
I wonder if the `_priv` part is really useful. You know you are operating on an xpriv and the return type is a private key. Kinda related: I'd like to explore the possibility of putting the derive and conversion functions into a trait (not here though) to allow generic code over x-keys.

sanket1729 replied:
After #755, I think this should be just `to_priv`. As mentioned and discussed in several places, keys are not related to signature algorithms.

sgeisler commented on src/util/bip32.rs:
```diff
@@ -498,11 +499,26 @@ impl ExtendedPrivKey {
+
+    /// Constructs BIP340 keypair for Schnorr signatures and Taproot use matching the internal
+    /// secret key representation.
+    pub fn to_schnorr_keypair<C: secp256k1::Signing>(&self, secp: &Secp256k1<C>) -> schnorr::KeyPair {
```
Same here `_keypair` seems redundant.

sgeisler commented on src/util/bip32.rs:
```diff
@@ -603,17 +619,31 @@ impl ExtendedPrivKey {
     }
 
+    /// Constructs ECDSA compressed public key matching internal public key representation.
+    pub fn to_ecdsa_pub(&self) -> ecdsa::PublicKey {
```
redundant `_pub`

sanket1729 replied:
Same, this should be `to_pub`

sgeisler commented on src/util/bip32.rs:
```diff
@@ -603,17 +619,31 @@ impl ExtendedPrivKey {
+
+    /// Constructs BIP340 x-only public key for BIP-340 signatures and Taproot use matching
+    /// the internal public key representation.
+    pub fn to_schnorr_pub(&self) -> schnorr::PublicKey {
```
redundant `_pub`

sanket1729 replied:
And this should be `to_x_only_pub`

sgeisler reviewed (dismissed):
Looks pretty good. I just now saw @sanket1729's comment :laughing: so take my comments with a grain of salt. As explained in the first comment my aversion to the more descriptive names comes from the idea of unifying the interface of x-keys via a common trait. For that the different names would be very annoying (but no blocker here, won't cut a release too soon anyway). I'll try to build it in a separate PR, the basic idea is to have:

```rust
pub trait XKey: Sized {
    type EcdsaKey;
    type SchnorrKey;

    fn derive<P: IntoIterator<Item=ChildNumber>>(&self, path: P) -> Self;
    fn to_ecdsa(&self) -> Self::EcdsaKey;
    fn to_schnorr(&self) -> Self::SchnorrKey;
}
```

apoelstra commented:
What is the status of this? Does it need a rebase and/or the PR description updated to have different commit IDs?

dr-orlovsky commented:
@apoelstra rebased, description update, ready for (re)reviews

sanket1729 reviewed (commented):
I think we should resolve our discussion https://github.com/rust-bitcoin/rust-bitcoin/issues/588#issuecomment-890586744 before proceeding with this.

dr-orlovsky commented:
@sanket1729 that discussion much more relates to previous PR #589 which is already merged (and a bitcoin-rust version with it is already released). If we need to change that approach, that should be a matter of a new PR.

This PR is very specifically removes unnecessary non-Secp256k1 pubkeys from BIP32/PSBT related data structures and does not directly related to the key type naming discussion within rust-bitcoin

sanket1729 commented:
@dr-orlovsky, there is a discussion to possibly undo things in #589 (in a non-breaking way). And there seems to be some support for that. If there is a consensus for that, this PR would be updated, namely, the function names that use `ecdsa_` and `schnorr_` should be `ec_` and `xonly_`(`x_only_`)

dr-orlovsky commented:
Why we can't do that alltogether after merging this PR, independently from it?

sanket1729 commented:
I don't have a strong opinion but appears that there is consensus for removing signature names from functions names using `x_only`, `ec`. 

This PR is not dependant on it and can get in without the other one, but there is no reason to have poor API names here just so we can change them later.

dr-orlovsky commented:
@sanket1729 can you pls explain what exactly you propose to update in this PR?

sanket1729 commented:
I was about to suggest APIs like 
```diff
- pub fn to_ecdsa_priv(&self) -> ecdsa::PrivateKey
+ pub fn to_ec_priv(&self) -> ecdsa::PrivateKey
```
but reading this seems odd that API when the return type is odd. 
ACK 7c565ec7a65e423756c42fc89b28a4fa7fccf80e.

We can change this later when we address #588

sanket1729 reviewed (dismissed):
ACK 7c565ec . Once we rename `ecdsa` and `schnorr` keys to `ec` and `xonly`, we should update the API names in this PR too.

apoelstra commented on examples/bip32.rs:
```diff
@@ -58,7 +58,7 @@ fn main() {
                          .unwrap()
                          .public_key;
-    let address = Address::p2wpkh(&public_key, network).unwrap();
+    let address = Address::p2wpkh(&ecdsa::PublicKey { compressed: true, key: public_key }, network).unwrap();
```
I'm a little tempted to just impl `From<secp256k1::PublicKey>` for `ecdsa::PublicKey` which does this.

thomaseizinger replied:
There is already a `new` function that defaults to compressed IIRC.

apoelstra commented on src/util/bip32.rs:
```diff
@@ -507,6 +507,21 @@ impl ExtendedPrivKey {
+    /// Constructs BIP340 keypair for Schnorr signatures and Taproot use matching the internal
+    /// secret key representation.
+    pub fn to_schnorr_keypair<C: secp256k1::Signing>(&self, secp: &Secp256k1<C>) -> schnorr::KeyPair {
+        schnorr::KeyPair::from_seckey_slice(secp, &self.private_key[..]).expect("BIP32 internal private key representation is broken")
```
Can you add a unit test which tries to parse an xpriv with a secret key of all zeros and one where the secret key is all `ff`s? I'm worried that this panic can be triggered.

dr-orlovsky replied:
Well, you can't construct xpriv with such secret keys - it errors during the construction on incorrect secret key. However I added general test case for this method.

GeneFerneau commented:
utACK [7c565ec](https://github.com/rust-bitcoin/rust-bitcoin/pull/590/commits/7c565ec7a65e423756c42fc89b28a4fa7fccf80e)

Kixunil commented on src/util/bip32.rs:
```diff
@@ -849,7 +878,7 @@ mod tests {
 
         let mut sk = ExtendedPrivKey::new_master(network, seed).unwrap();
-        let mut pk = ExtendedPubKey::from_private(secp, &sk);
+        let mut pk = ExtendedPubKey::from_priv(secp, &sk);
```
Is this used so often that removing three letters is worth readability and larger diff?

dr-orlovsky replied:
It was not removed for readability, but for standardization. Most of method names uses `pub`/`priv` or `secret`, not `private`

dr-orlovsky commented:
@apoelstra I added the test in the last commit (by amending it). Unfortunately it required to change crate policy from `forbit(unsafe_code)` to `deny(unsafe_code)` but only in test configuration. The function does not panics but errors.

apoelstra commented:
@dr-orlovsky that's cool, but I think you're misunderstanding the test that I want. What you've done (involving unsafe code) seems to just test rust-secp functions .... somewhich which not only belongs in rust-secp, but which doesn't require unsafety even if you were testing it here.

What I would like is a stringified xpriv with a bad privkey embedded in it, and an attempt to parse it into a xpriv struct.

dr-orlovsky commented:
@apoelstra done

apoelstra commented on src/util/bip32.rs:
```diff
@@ -1080,5 +1111,50 @@ mod tests {
+    }
+
+    #[test]
+    #[should_panic(expected = "Secp256k1(InvalidSecretKey)")]
```
In ca4e90cd3eae74ad6f7a93af22aacf3eec53e73d:

I have a mild preference for using `unwrap_err` (and then maybe `to_string`) to compare errors inside the test, rather than using `should_panic`.

apoelstra commented on src/util/bip32.rs:
```diff
@@ -1080,5 +1111,50 @@ mod tests {
+    #[should_panic(expected = "Secp256k1(InvalidSecretKey)")]
+    fn schnorr_broken_privkey_ffs() {
+        // Xpriv having secret key set to all 0xFF's
+        let xpriv_str = "xprv9s21ZrQH143K24Mfq5zL5MhWK9hUhhGbd45hLXo2Pq2oqzMMo63oStZzFAzHGBP2UuGCqWLTAPLcMtD9y5gkZ6Eq3Rjuahrv17fENZ3QzxW";
```
In ca4e90cd3eae74ad6f7a93af22aacf3eec53e73d:

Note to reviewers: this has a chaincode of all zeros but the secret key is all `ff`s.

apoelstra reviewed (approved):
ACK ca4e90cd3eae74ad6f7a93af22aacf3eec53e73d

sanket1729 commented on src/util/bip32.rs:
```diff
@@ -612,7 +627,7 @@ impl ExtendedPrivKey {
 impl ExtendedPubKey {
     /// Derives a public key from a private key
-    pub fn from_private<C: secp256k1::Signing>(secp: &Secp256k1<C>, sk: &ExtendedPrivKey) -> ExtendedPubKey {
+    pub fn from_priv<C: secp256k1::Signing>(secp: &Secp256k1<C>, sk: &ExtendedPrivKey) -> ExtendedPubKey {
```
I am not a big fan of this change. But I can live with this. This basically brings a compile error downstream that is difficult to solve. I think we should at least add this in release notes. 

The user at least has some guidance from the compiler when we change the function signature while keeping the name the same. In this case, it would be just say function not found.

sanket1729 reviewed (commented):
If we can rebase this after #755, it would be great. I think the naming can improve substantially then.

dr-orlovsky commented:
@sanket1729 addressed your review and re-based. From what I see from the failed test in the last commit we need another point release for secp:

```
binary operation `==` cannot be applied to type `secp256k1::KeyPair`
```

I was working before with the previous secp version.

Kixunil commented on src/util/bip32.rs:
```diff
@@ -611,17 +627,37 @@ impl ExtendedPrivKey {
+            compressed: true,
+            key: self.public_key
+        }
+    }
```
This method feels more like a getter because it loses some information (conversion like `to_` usually don't) . Getters in Rust don't have prefixes (`get_` or `to_`), so the name should be `public_key()` I think.

Kixunil commented on src/util/bip32.rs:
```diff
@@ -611,17 +627,37 @@ impl ExtendedPrivKey {
+    /// the internal public key representation.
+    pub fn to_x_only_pub(&self) -> XOnlyPublicKey {
+        XOnlyPublicKey::from(self.public_key)
+    }
```
This conversion looks wrong. If I understand this correctly it may change the parity of the key making it effectively different than the original one. I can't imagine how it'd be a good idea to do that and even if it has some use cases, having it here publicly could give people wrong ideas. Please explain if I'm missing something.

FWIW I'm not even sure if `impl From<PublicKey> for XOnlyPublicKey` was a good idea given it loses one bit of information.

Also, same issue with getter as above.

Kixunil reviewed (changes requested):
The strange conversion is too worrying to me to ACK this, please explain or remove.

(side note: abbreviating `private` to `priv` hurts my eyes but not going to argue about it)

dr-orlovsky commented:
@Kixunil it is a part of the standard to drop first byte in public key - to x-only public key conversion for BIP32: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#public-key-conversion

As regarding `priv` vs `private`, I changed the name to follow already used naming of other methods. I do not mind renaming all methods from `priv` to `private` but this will be a larger breaking change.

apoelstra commented:
I think @dr-orlovsky is correct. I agree, it's weird to drop a bit of information in a "conversion" but this bit of information is not used in signature validation, and this is the standard way to obtain an x-only key from an ordinary one.

Kixunil reviewed (dismissed):
ACK 1ff5730d21bad940d18b975199b2c0a393056644

sanket1729 commented on src/util/bip32.rs:
```diff
@@ -1116,5 +1117,50 @@ mod tests {
+        let secp = Secp256k1::new();
+        let xpriv = ExtendedPrivKey::from_str("xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi").unwrap();
+        let pair = KeyPair::from_seckey_slice(&secp, &xpriv.private_key[..]).unwrap();
+        assert_eq!(xpriv.to_schnorr_keypair(&secp), pair);
```
I don't see any value in this test? What is this test testing? It is a one-line function code that is repeated in the test. 

If you feel there is value in this, we can check that the x-only public keys from the keypairs are the same for now. After another point release, we can re-add this test. I don't think we are losing any features for the next release because of this.

sanket1729 reviewed (commented):
Left two comments. I don't think we should wait for a point release of secp for a purpose of a test (which I think is not testing anything concrete). We can add the test later after a point release of secp.

dr-orlovsky commented:
@sanket1729 addressed issues and removed test.

@Kixunil pls re-ACK

Kixunil commented:
Would've reacked but it's uncompilable...

dr-orlovsky commented:
@Kixunil sorry, my bad: introduced last minute "improvement" with additional commit and didn't properly checked it. Should be fixed now

Kixunil commented:
Still some issue.

dr-orlovsky commented:
@Kixunil I do not know what the ... is happening with CI, but it perfectly builds locally while it fails with exactly the same config on GitHub:

```
orlovsky@alienware:~/repos/rust-bitcoin$ rustup update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: cleaning up downloads & tmp directories
orlovsky@alienware:~/repos/rust-bitcoin$ cargo +stable build --features=no-std --no-default-features
    Finished dev [unoptimized + debuginfo] target(s) in 0.27s
orlovsky@alienware:~/repos/rust-bitcoin$ cargo +nightly build --features=no-std --no-default-features
    Finished dev [unoptimized + debuginfo] target(s) in 0.26s
```

sanket1729 commented:
@dr-orlovsky , the tests are not building. Not the binary.

sanket1729 commented:
This probably has something to do with `DisplaySecret` PR in rust-secp256k1. I think you are probably the best person to debug this since you authored it :P

sanket1729 commented on src/util/bip32.rs:
```diff
@@ -44,7 +45,8 @@ impl_array_newtype!(Fingerprint, u8, 4);
 /// Extended private key
-#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
```
Why are you doing this? I think this is causing the issue? Why cannot use derive debug in no-std? And how is that related to this PR.

dr-orlovsky replied:
We need this since `SecretKey` in the upstream does custom `Debug` which works only on `std`. Without this change it foes not compile on `no_std`

sanket1729 commented:
I think we should leave all the secret display stuff out of this PR and this should build.

dr-orlovsky commented:
@Kixunil @sanket1729 sorry it was my bad: probably was too tired yesterday and thought that I had no-std `Debug` impl for `ExtendedPrivateKey`. Fixed

apoelstra reviewed (approved):
ACK cf0c48cc86b1f9710588987db8137362275166ad

It's a little eyebrow-raising that the nostd `Debug` impl just silently omits the secret key, but whatever. I have no opinions about the renaming, though I think after we get all the key-changing PRs in (but before we do the Taproot release) we should go through everything and possibly do a rename cleanup.

dr-orlovsky commented:
@apoelstra there is not a lot of options for `no_std` `Debug`: in `secp256k1` secret key do not have any `Debug` in `no_std` and is hashed in `std`. If we just plainly print secret key here we will abandon "never display secrets in logs" policy we introduced in `secp256k1` with that hashing.

apoelstra commented:
Right, I understand that we have no good options here. I think I might've stuck the literal string `[secret]` in there or something, but just eliding the field is fine.

sanket1729 commented on src/util/bip32.rs:
```diff
@@ -638,6 +659,20 @@ impl ExtendedPubKey {
 
+    /// Constructs ECDSA compressed public key matching internal public key representation.
+    pub fn to_pub(&self) -> ecdsa::PublicKey {
+        ecdsa::PublicKey {
```
Note that there would be some conflicts with #757

sanket1729 reviewed (approved):
ACK cf0c48cc86b1f9710588987db8137362275166ad. #757 might need rework after this
